airflow/providers/hashicorp/hooks/vault.py

```diff
@@ -0,0 +1,310 @@
+           Mutually exclusive with gcp_keyfile_dict
+    :type gcp_key_path: str
+    :param gcp_keyfile_dict: Dictionary of keyfile parameters. (for ``gcp`` auth_type).
+           Mutually exclusive with gcp_key_path
+    :type gcp_keyfile_dict: dict
+    :param gcp_scopes: Comma-separated string containing GCP scopes (for ``gcp`` auth_type)
+    :type gcp_scopes: str
+    :param azure_tenant_id: Tenant of azure (for ``azure`` auth_type)
+    :type azure_tenant_id: str
+    :param azure_resource: Resource if of azure (for ``azure`` auth_type)
+    :type azure_resource: str
+    :param radius_host: Host for radius (for ``radius`` auth_type)
+    :type radius_host: str
+    :param radius_port: Port for radius (for ``radius`` auth_type)
+    :type radius_port: int
+
+    """
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        vault_conn_id: str,
+        auth_type: Optional[str] = None,
+        mount_point: Optional[str] = None,
+        kv_engine_version: Optional[int] = None,
+        token: Optional[str] = None,
+        role_id: Optional[str] = None,
+        kubernetes_role: Optional[str] = None,
+        kubernetes_jwt_path: Optional[str] = None,
+        gcp_key_path: Optional[str] = None,
+        gcp_keyfile_dict: Optional[dict] = None,
+        gcp_scopes: Optional[str] = None,
+        azure_tenant_id: Optional[str] = None,
+        azure_resource: Optional[str] = None,
+        radius_host: Optional[str] = None,
+        radius_port: Optional[int] = None
+    ):
+        super().__init__()
+        self.connection = self.get_connection(vault_conn_id)
+
+        if not auth_type:
+            auth_type = self.connection.extra_dejson.get('auth_type')
+        if not auth_type:
+            auth_type = "token"
+
+        if not mount_point:
+            mount_point = self.connection.extra_dejson.get('mount_point')
+
+        if not kv_engine_version:
+            conn_version = self.connection.extra_dejson.get("kv_engine_version")
+            try:
+                kv_engine_version = int(conn_version) if conn_version else DEFAULT_KV_ENGINE_VERSION
+            except ValueError:
+                raise VaultError(f"The version is not an int: {conn_version}. ")
+
+        if not role_id:
+            role_id = self.connection.extra_dejson.get('role_id')
+
+        if not token:
+            token = self.connection.extra_dejson.get('token')
+
+        azure_resource, azure_tenant_id = \
+            self._get_azure_parameters_from_connection(azure_resource, azure_tenant_id)
+        gcp_key_path, gcp_keyfile_dict, gcp_scopes = \
+            self._get_gcp_parameters_from_connection(gcp_key_path, gcp_keyfile_dict, gcp_scopes)
+        kubernetes_jwt_path, kubernetes_role = \
+            self._get_kubernetes_parameters_from_connection(kubernetes_jwt_path, kubernetes_role)
+        radius_host, radius_port = self._get_radius_parameters_from_connection(radius_host, radius_port)
```

Comment:
These methods would be called for all auth_types i.e. even if the auth is just token, it would call all these functions and try to get values from Connection

Reply:
Yep. I did it deliberately, it's pretty harmless (as these are just parameter retrievals and make the logic slightly simpler. I can make slight change and make it conditional though - no problem.

Reply:
Fixed and rebased. I also added a few more tests and separated the "missing" tests into separate exception to be very clearly about the reason. See the latest fixup.